Expand Supports syntax (part deux)

kbrock opened the pull request:
Part of project: https://github.com/ManageIQ/manageiq/issues/21179

Also merge: https://github.com/ManageIQ/manageiq-ui-classic/pull/8366

This is based upon a conversation started by @chessbyte : #21269 which started in [vmware-provider](https://github.com/ManageIQ/manageiq-providers-vmware/pull/723#discussion_r650005096)


It came up a few times and now that we have reworked supports a bit, it seemed like a good time to see revisit and see if this new syntax makes more sense.

## Ideas in discussion

### Before

```ruby
    supports :check_compliance
    supports_not :check_compliance
    supports_not :check_compliance, :reason => "Feature not available/supported"

    supports :check_compliance do
      unsupported_reason_add(:check_compliance, "No compliance policies assigned") unless has_compliance_policies?
    end
```

Most of the `supports_not` blocks have been transitioned to use the default reason:

|  #  |comment|
|----:|:------|
| 245 | `supports :feature`|
| 151 | `supports_not :feature`|
|  23 | `supports_not :feature, :reason => ""`|
| 169 | `supports :feature do ; end`|

### String Syntax

```ruby
    supports :check_compliance do
      "No compliance policies assigned" unless has_compliance_policies?
    end
```

This gives the end user more of an understanding as to why this operation is not available.
Half of the time this is an `if`/`elsif` chain.
Returning `nil` means that this feature is supported.

Of the reasons given, a quarter could just use the default value:

| %  |comment|
|---:|:------|
| 50 | multiple reasons per block|
| 25 | nondescript single reason |
| 25 | single reason|

### Boolean Syntax

We have dropped the use of returning a boolean. Most of those cases have been converted to the standard `supports` / `supports_not` syntax.

kbrock commented:
extra note:

I am wondering if we want a chain method, to base the supports either upon a parent feature or upon another feature. host, vm, and vm_or_template all base one supports upon a core feature.

```ruby
supports :stop do
  if !supports?(:control)
    unsupported_reason(:control)
  elsif !vm_powered_on?
    _('The VM is not powered on')
  end
end
#
chain_supports(:stop, :control) { _('The VM is not powered on') unless vm_powered_on? }
#
supports(:stop => :control) { _('The VM is not powered on') unless vm_powered_on? }
```

The two are more compact, but they don't necessarily read better.
Also, they can depend upon 2. My imagined code behind this feels unwieldy, and I don't think it makes the code base any more readable

kbrock commented:
Implementing the boolean or string logic looks relatively straight forward:

```ruby
if unsupported.key?(feature).nil? && result == false || result.kind_of?(String)
  unsupported_reason_add(feature, reason_or_default(result))
end
```

kbrock commented:
cops did not update to new code
I'm finding that just using this for a day has given me new ideas to simplify the implementation.

push(es):
- fixed use case for strings
- added boolean case (and re-pushed a few times)

kbrock commented:
kicking cops

kbrock commented:
WIP:
- @Fryguy and possibly some others want to give this a little time to come up with better ideas

kbrock commented on app/models/host.rb:
```diff
@@ -198,35 +198,33 @@ class Host < ApplicationRecord
 
   supports :start do
-    valid_states = respond_to?(:vim_power_up_from_standby) ? %w[off standby] : "off"
+    valid_states = respond_to?(:vim_power_up_from_standby) ? %w[off standby] : %w[off]
```
as a followup, will move this into vmware

For now, want to just focus on one repo at a time.

kbrock replied:
https://github.com/ManageIQ/manageiq-providers-vmware/pull/809

Fryguy commented:
Ok, sorry it took me some time to formulate this, but my fundamental problem with the approach is what you actually already laid out in the Implementation section of the OP.  The return value from the block has multiple truthiness values that are contradictory and it's confusing to read.  I see you wrote that it's not, but to me it's very confusing.

---

Here's what I had started writing before reading your comment there, but it's a similar table:

Given the word `supports` is a positive word in contrast to `supports_not` which I consider negative, and truthiness-wise I consider true to be a positive word and false to be a negative word, you get this:

| code | positivity | supported? |
| --- | --- | --- |
| supports { true } | positive { positive } | yes |
| support { "xxx" } | postive { positive } |  no |
| supports { false } | positive { negative } | no |
| supports { nil } | positive { negative } | yes |

---

I'll leave it up to @agrare to decide though, since this code falls almost entirely in the providers space.

kbrock commented:
I almost did not introduce the `true` vs `false`, concept because it seemed too complicated. You didn't like the way I described the issue and wrote it out again. (the very fact that you did this shows there this concept is confusing)

But in the end of the day, the most important thing is to read the models and decide if it is clear.

For me, the new way was easy to read, concise, and even mixing strings and booleans were clear.

chessbyte commented on app/models/host.rb:
```diff
@@ -198,35 +198,33 @@ class Host < ApplicationRecord
-  supports :reset do
-    unsupported_reason_add(:reset, unsupported_reason(:ipmi)) unless supports?(:ipmi)
-  end
+  supports(:reset) { unsupported_reason(:ipmi) }
```
was the `unless supports?(:ipmi)` dropped intentionally?

kbrock replied:
Yea. Once looking into this feature it started morphing on its own.
Love it when this happens.

```ruby
unsupported_reason_add(:reset, unsupported_reason(:ipmi)) unless supports?(:ipmi)
## =>
unsupported_reason(:ipmi) unless supports?(:ipmi)
## =>
if supports?(:ipmi)
  nil
else
  unsupported_reason(:ipmi)
end
## =>
if supports?(:ipmi)
  unsupported_reason(:ipmi) # nil
else
  unsupported_reason(:ipmi)
end
## =>
 unsupported_reason(:ipmi)
```

kbrock commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -187,7 +187,12 @@ def define_supports_feature_methods(feature, is_supported: true, reason: nil, &b
           if block_given?
             begin
-              instance_eval(&block)
+              result = instance_eval(&block)
+              # if no errors yet but result was false or an error message
+              # then add the error
+              if !unsupported.key?(feature) && (result.kind_of?(String) || result == false)
+                unsupported_reason_add(feature, SupportsFeatureMixin.reason_or_default(result))
+              end
```
@chessbyte I was pretty psyched when the end solution was this simple. (along the way it was more complicated)

kbrock commented on app/models/manageiq/providers/container_manager.rb:
```diff
@@ -56,11 +56,7 @@ class ContainerManager < BaseManager
-        unsupported_reason_add(:external_logging, _('This provider type does not support external_logging'))
-      end
-    end
+    supports(:external_logging) { respond_to?(:external_logging_route_name) }
```
convert to `supports`/`supports_not`

kbrock commented:
update:

- rebased
- made more cases single liners
- fixed syntax errors (for making single liners without parens)
- removed boolean option (most just became simple `supports_not :feature ; supports :feature`

kbrock commented on app/models/metric/ci_mixin.rb:
```diff
@@ -23,7 +23,7 @@ module Metric::CiMixin
       metrics_capture_klass = "#{self.class.module_parent.name}::MetricsCapture".safe_constantize
       unless metrics_capture_klass&.method_defined?(:perf_collect_metrics)
-        unsupported_reason_add(:capture, _('This provider does not support metrics collection'))
+        _('This provider does not support metrics collection')
```
convert to `supports :capture` in the subclasses

kbrock replied:
moved to https://github.com/ManageIQ/manageiq/issues/21989

kbrock commented on app/models/storage.rb:
```diff
@@ -71,17 +71,13 @@ class Storage < ApplicationRecord
   supports :smartstate_analysis do
     if !ext_management_system&.class&.supports?(:smartstate_analysis)
-      unsupported_reason_add(:smartstate_analysis, _("Smartstate Analysis cannot be performed on selected Datastore"))
+      _("Smartstate Analysis cannot be performed on selected Datastore")
```
this also feels like this should be part of the underlying class

kbrock replied:
added to https://github.com/ManageIQ/manageiq/issues/21989

kbrock commented:
kicking (strange 2.6 hang)

Fryguy commented on app/models/service_template.rb:
```diff
@@ -72,8 +72,11 @@ class ServiceTemplate < ApplicationRecord
-    unsupported_reason_add(:order, 'Service template does not belong to a service catalog') unless service_template_catalog
-    unsupported_reason_add(:order, 'Service template is not configured to be displayed') unless display
+    if !service_template_catalog
+      'Service template does not belong to a service catalog'
+    elsif !display
+      'Service template is not configured to be displayed'
```
Curious this is not i18n'd...we can do that in a follow up.

kbrock replied:
we have a lot of these that are not internationalized''

kbrock commented:
update:
- rebased to fix build failure
- moved TODOs to separate issue https://github.com/ManageIQ/manageiq/issues/21989

kbrock commented:
update:
- rebase to fix merge conflict

kbrock commented:
We have dropped the case for boolean.
I decided to include the text here for future reference

### REDACTED: true false syntax

```ruby
    supports(:check_compliance:) { has_compliance_policies? }
```

true returned means that this DOES support compliance
false return means that this does not support compliance.

#### REDACTED: Implementation

I added this as the last commit. The implementation is simple enough, but it is a little ambiguous/confusing if you think hard enough (too hard?).

| block return value | logical value | supported? |
|:-------|:------|:----|
| true   | true  | yes |
| false  | false | no  |
| String | true  | no  |
| nil    | false | yes |

The fact that a logical true can mean supported or not is a little confusing, and makes this option a little error prone. But in practice the code reads well enough.

Introducing the boolean did simplify a few in core, and many more in the associated providers.

#### Current implementation

As mentioned above, it was simpler to just remove the need for these cases.

kbrock commented:
anyway
TL;DR: currently only change is it allows you to return a string rather than calling `unsupported_reason_add`

agrare commented:
@miq-bot cross-repo-tests /all

Fryguy reviewed (approved):
Fantastic work!

kbrock commented:
Thanks for the x-repo @agrare 
It looks like the failure is the ttl issue with the API.

Can we merge this with that sporadic failure?

agrare commented:
@kbrock yeah API looks sporadic and ui-service failure looks unrelated (npm) but the ui-classic failure looks like it might be related https://github.com/ManageIQ/manageiq-cross_repo-tests/runs/7414529833?check_suite_focus=true#step:6:2847

kbrock commented:
update:
- swapped order of snapshot checks, same answers, but messages are a little different
- note: still using default error message when snapshot is not supported (vs custom error before)

kbrock commented:
@miq-bot cross-repo-tests https://github.com/ManageIQ/manageiq-ui-classic/pull/8366

chessbyte commented:
Glad to see this over the finish line!
